qiskit/algorithms/minimum_eigensolvers/qaoa.py

```diff
@@ -0,0 +1,137 @@
+        [3]: Hadfield, S. et al, "From the Quantum Approximate Optimization Algorithm to a Quantum
+            Alternating Operator Ansatz" `arXiv:1709.03489 <https://arxiv.org/abs/1709.03489>`__
+        [4]: Egger, D. J., Marecek, J., Woerner, S., "Warm-starting quantum optimization"
+            `arXiv: 2009.10095 <https://arxiv.org/abs/2009.10095>`__
+    """
+
+    def __init__(
+        self,
+        sampler: BaseSampler,
+        optimizer: Optimizer | Minimizer,
+        *,
+        reps: int = 1,
+        initial_state: QuantumCircuit | None = None,
+        mixer: QuantumCircuit | BaseOperator | PauliSumOp = None,
+        initial_point: np.ndarray | None = None,
+        aggregation: float | Callable[[list[float]], float] | None = None,
+        callback: Callable[[int, np.ndarray, float, dict[str, Any]], None] | None = None,
+    ) -> None:
+        """
+        Args:
+            sampler: The sampler primitive to sample the circuits.
+            optimizer: A classical optimizer, see also :class:`~qiskit.algorithms.VQE` for
+                more details on the possible types.
+            reps: the integer parameter :math:`p` as specified in https://arxiv.org/abs/1411.4028,
+                Has a minimum valid value of 1.
+            initial_state: An optional initial state to prepend the QAOA circuit with.
+            mixer: the mixer Hamiltonian to evolve with or a custom quantum circuit. Allows support
+                of optimizations in constrained subspaces as per https://arxiv.org/abs/1709.03489
+                as well as warm-starting the optimization as introduced
+                in http://arxiv.org/abs/2009.10095.
+            initial_point: An optional initial point (i.e. initial parameter values)
+                for the optimizer. If ``None`` then it will simply compute a random one.
+            aggregation: A float or callable to specify how the objective function evaluated on the
+                basis states should be aggregated.
+            callback: A callback that can access the intermediate data at each optimization step.
+                These data are: the evaluation count, the optimizer parameters for the ansatz, the
+                evaluated value, the the metadata dictionary.
+        """
+        validate_min("reps", reps, 1)
+
+        self.reps = reps
+        self.mixer = mixer
+        self.initial_state = initial_state
+        self._cost_operator = None
+
+        super().__init__(
+            sampler=sampler,
+            ansatz=None,
+            optimizer=optimizer,
+            initial_point=initial_point,
+            aggregation=aggregation,
+            callback=callback,
+        )
+
+    def _check_operator_ansatz(self, operator: BaseOperator | PauliSumOp):
+        # Recreates a circuit based on operator parameter.
+        self.ansatz = QAOAAnsatz(
+            operator, self.reps, initial_state=self.initial_state, mixer_operator=self.mixer
```

Comment:
QAOA ansatz takes OperatorBase as an operator, so while PauliSumOp is ok, which is normally what we have, I think BaseOperator (i.e quantum info Operator support) will need EvolvedOperatorAnsatz looking at.